comp/core/tagger/taggerimpl/subscriber/subscription_manager.go

```diff
@@ -0,0 +1,172 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package subscriber
+
+import (
+	"sync"
+
+	"github.com/DataDog/datadog-agent/comp/core/tagger/telemetry"
+	"github.com/DataDog/datadog-agent/comp/core/tagger/types"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// SubscriptionManager allows processes to subscribe to entity events generated from a
+// tagger.
+type SubscriptionManager interface {
+	Subscribe(id string, filter *types.Filter, events []types.EntityEvent) types.Subscription
+	Unsubscribe(subscriptionID string)
+	Notify(events []types.EntityEvent)
+}
+
+// subscriptionManager implements SubscriptionManager
+type subscriptionManager struct {
+	sync.RWMutex
+	subscribers    map[string]*Subscriber
+	prefixToSub    map[types.EntityIDPrefix][]*Subscriber
+	telemetryStore *telemetry.Store
+}
+
+// NewSubscriptionManager creates and returns a new subscription manager
+func NewSubscriptionManager(telemetryStore *telemetry.Store) SubscriptionManager {
+	return &subscriptionManager{
+		subscribers:    make(map[string]*Subscriber),
+		prefixToSub:    make(map[types.EntityIDPrefix][]*Subscriber),
+		telemetryStore: telemetryStore,
+	}
+}
+
+// Subscribe returns a channel that receives a slice of events whenever an
+// entity is added, modified or deleted. It can send an initial burst of events
+// only to the new subscriber, without notifying all of the others.
+func (sm *subscriptionManager) Subscribe(id string, filter *types.Filter, events []types.EntityEvent) types.Subscription {
+	// this is a `ch []EntityEvent` instead of a `ch EntityEvent` to
+	// improve throughput, as bursts of events are as likely to occur as
+	// isolated events, especially at startup or with collectors that
+	// periodically pull changes.
+	ch := make(chan []types.EntityEvent, bufferSize)
+	subscriber := &Subscriber{
+		filter:  *filter,
+		id:      id,
+		ch:      ch,
+		manager: sm,
+	}
+
+	sm.Lock()
+	sm.subscribers[id] = subscriber
```

Comment:
should add a check to know it the ID is already used? 
In that case should we return an error? and maybe change the function signature to `Subscribe(id string, filter *types.Filter, events []types.EntityEvent) (types.Subscription, error)` 

So like this you can also handle the case of the `noopTagger` implementation by returning an error that say "not implemented"

Reply:
Agreed 👍 

Will update it